spacy/pipeline/spancat_exclusive.py

```diff
@@ -0,0 +1,308 @@
+window_size = 1
+maxout_pieces = 3
+depth = 4
+"""
+
+DEFAULT_SPANCAT_MODEL = Config().from_str(spancat_exclusive_default_config)["model"]
+
+
+@Language.factory(
+    "spancat_exclusive",
+    assigns=["doc.spans"],
+    default_config={
+        "spans_key": "sc",
+        "model": DEFAULT_SPANCAT_MODEL,
+        "suggester": {"@misc": "spacy.ngram_suggester.v1", "sizes": [1, 2, 3]},
+        "scorer": {"@scorers": "spacy.spancat_scorer.v1"},
+    },
+    default_score_weights={"spans_sc_f": 1.0, "spans_sc_p": 0.0, "spans_sc_r": 0.0},
+)
+def make_spancat(
+    nlp: Language,
+    name: str,
+    suggester: Suggester,
+    model: Model[Tuple[List[Doc], Ragged], Floats2d],
+    spans_key: str,
+    scorer: Optional[Callable],
+    negative_weight: float = 1.0,
+    allow_overlap: bool = True,
+) -> "Exclusive_SpanCategorizer":
+    """Create a SpanCategorizerExclusive component. The span categorizer consists of two
+    parts: a suggester function that proposes candidate spans, and a labeller
+    model that predicts a single label for each span.
+
+    suggester (Callable[[Iterable[Doc], Optional[Ops]], Ragged]): A function that suggests spans.
+        Spans are returned as a ragged array with two integer columns, for the
+        start and end positions.
+    model (Model[Tuple[List[Doc], Ragged], Floats2d]): A model instance that
+        is given a list of documents and (start, end) indices representing
+        candidate span offsets. The model predicts a probability for each category
+        for each span.
+    spans_key (str): Key of the doc.spans dict to save the spans under. During
+        initialization and training, the component will look for spans on the
+        reference document under the same key.
+    scorer (Optional[Callable]): The scoring method. Defaults to
+        Scorer.score_spans for the Doc.spans[spans_key] with overlapping
+        spans allowed.
+    negative_weight (float): Multiplier for the loss terms.
+        Can be used to down weigh the negative samples if there are too many.
+    allow_overlap (bool): If True the data is assumed to
+        contain overlapping spans.
+    """
+    return Exclusive_SpanCategorizer(
+        nlp.vocab,
+        suggester=suggester,
+        model=model,
+        spans_key=spans_key,
+        negative_weight=negative_weight,
+        name=name,
+        scorer=scorer,
+        allow_overlap=allow_overlap,
+    )
+
+
+@dataclass
+class Ranges:
+    """
+    Helper class help avoid storing overlapping span.
+    """
+
+    def __init__(self):
+        self.ranges = set()
+
+    def add(self, i, j):
+        for e in range(i, j):
+            self.ranges.add(e)
+
+    def __contains__(self, rang):
+        i, j = rang
+        for e in range(i, j):
+            if e in self.ranges:
+                return True
+        return False
+
+
+class Exclusive_SpanCategorizer(SpanCategorizer):
+    """Pipeline component to label non-overlapping spans of text.
```

Comment:
Is "non-overlapping" a correct description? (This doesn't make sense in combination with the `allow_overlap` option in particular.)

Reply:
Updating to: "Pipeline component to label spans with exclusive classes"